File: ontrack-extension-jenkins/src/test/java/net/nemerosa/ontrack/extension/jenkins/model/JenkinsJobPropertyTypeTest.java
package net.nemerosa.ontrack.extension.jenkins.model;

import com.fasterxml.jackson.databind.JsonNode;
import net.nemerosa.ontrack.extension.indicators.IndicatorsExtensionFeature;
import net.nemerosa.ontrack.extension.jenkins.*;
import net.nemerosa.ontrack.extension.scm.SCMExtensionFeature;
import net.nemerosa.ontrack.model.security.ProjectConfig;
import net.nemerosa.ontrack.model.security.SecurityService;
import net.nemerosa.ontrack.model.structure.*;
import org.junit.Before;
import org.junit.Test;

import java.util.Set;

import static net.nemerosa.ontrack.json.JsonUtils.object;
import static net.nemerosa.ontrack.test.TestUtils.assertJsonEquals;
import static org.junit.Assert.*;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class JenkinsJobPropertyTypeTest {

    private JenkinsJobPropertyType type;
    private SecurityService securityService;
    private JenkinsConfigurationService configurationService;

    @Before
    public void before() {
        securityService = mock(SecurityService.class);
        configurationService = mock(JenkinsConfigurationService.class);
        type = new JenkinsJobPropertyType(
                new JenkinsExtensionFeature(
                        new IndicatorsExtensionFeature(),
                        new SCMExtensionFeature()
                ),
                configurationService
        );
    }

    @Test
    public void applies() {
        Set<ProjectEntityType> types = type.getSupportedEntityTypes();
        assertTrue(types.contains(ProjectEntityType.PROJECT));
        assertTrue(types.contains(ProjectEntityType.BRANCH));
        assertTrue(types.contains(ProjectEntityType.PROMOTION_LEVEL));
        assertTrue(types.contains(ProjectEntityType.VALIDATION_STAMP));
        assertFalse(types.contains(ProjectEntityType.BUILD));
        assertFalse(types.contains(ProjectEntityType.PROMOTION_RUN));
        assertFalse(types.contains(ProjectEntityType.VALIDATION_RUN));
    }

    @Test
    public void canEdit() {
        Project p1 = Project.of(new NameDescription("P1", "Project 1")).withId(ID.of(1));
        Project p2 = Project.of(new NameDescription("P2", "Project 2")).withId(ID.of(2));
        Branch b1 = Branch.of(p1, new NameDescription("B1", "Branch 1")).withId(ID.of(10));
        Branch b2 = Branch.of(p2, new NameDescription("B2", "Branch 2")).withId(ID.of(20));
        when(securityService.isProjectFunctionGranted(1, ProjectConfig.class)).thenReturn(true);
        when(securityService.isProjectFunctionGranted(2, ProjectConfig.class)).thenReturn(false);
        assertTrue(type.canEdit(b1, securityService));
        assertFalse(type.canEdit(b2, securityService));
    }

    @Test
    public void canView() {
        Project p1 = Project.of(new NameDescription("P1", "Project 1")).withId(ID.of(1));
        Project p2 = Project.of(new NameDescription("P2", "Project 2")).withId(ID.of(2));
        Branch b1 = Branch.of(p1, new NameDescription("B1", "Branch 1")).withId(ID.of(10));
        Branch b2 = Branch.of(p2, new NameDescription("B2", "Branch 2")).withId(ID.of(20));
        assertTrue(type.canView(b1, securityService));
        assertTrue(type.canView(b2, securityService));
    }

    @Test
    public void forStorage() {
        JenkinsConfiguration configuration = new JenkinsConfiguration(
                "MyConfig",
                "http://jenkins",
                "user",
                "secret"
        );
        assertJsonEquals(
                object()
                        .with("configuration", "MyConfig")
                        .with("job", "MyJob")
                        .end(),
                type.forStorage(
                        new JenkinsJobProperty(
                                configuration,
                                "MyJob"
                        )
                )
        );
    }

    @Test
    public void fromStorage() {
        JenkinsConfiguration configuration = new JenkinsConfiguration(
                "MyConfig",
                "http://jenkins",
                "user",
                "secret"
        );
        when(configurationService.getConfiguration("MyConfig")).thenReturn(configuration);
        // Stored JSON
        JsonNode node = object()
                .with("configuration", "MyConfig")
                .with("job", "MyJob")
                .end();
        // Reading
        JenkinsJobProperty retrieved = type.fromStorage(node);
        assertEquals(configuration.getUrl(), retrieved.getConfiguration().getUrl());
        assertEquals("MyJob", retrieved.getJob());
    }

}

File: ontrack-extension-jenkins/src/test/java/net/nemerosa/ontrack/extension/jenkins/model/JenkinsJobPropertyTypeTest.kt
package net.nemerosa.ontrack.extension.jenkins.model

import com.fasterxml.jackson.databind.JsonNode
import io.mockk.every
import io.mockk.mockk
import net.nemerosa.ontrack.extension.indicators.IndicatorsExtensionFeature
import net.nemerosa.ontrack.extension.jenkins.*
import net.nemerosa.ontrack.extension.scm.SCMExtensionFeature
import net.nemerosa.ontrack.json.asJson
import net.nemerosa.ontrack.model.security.ProjectConfig
import net.nemerosa.ontrack.model.security.SecurityService
import net.nemerosa.ontrack.model.structure.Branch.Companion.of
import net.nemerosa.ontrack.model.structure.ID.Companion.of
import net.nemerosa.ontrack.model.structure.NameDescription
import net.nemerosa.ontrack.model.structure.Project.Companion.of
import net.nemerosa.ontrack.model.structure.ProjectEntityType
import net.nemerosa.ontrack.test.TestUtils
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class JenkinsJobPropertyTypeTest {

    private lateinit var type: JenkinsJobPropertyType
    private lateinit var securityService: SecurityService
    private lateinit var configurationService: JenkinsConfigurationService

    @BeforeEach
    fun before() {
        securityService = mockk<SecurityService>()
        configurationService = mockk<JenkinsConfigurationService>()
        type = JenkinsJobPropertyType(
            JenkinsExtensionFeature(
                IndicatorsExtensionFeature(),
                SCMExtensionFeature()
            ),
            configurationService
        )
    }

    @Test
    fun applies() {
        val types: Set<ProjectEntityType> = type.supportedEntityTypes
        assertTrue(types.contains(ProjectEntityType.PROJECT))
        assertTrue(types.contains(ProjectEntityType.BRANCH))
        assertTrue(types.contains(ProjectEntityType.PROMOTION_LEVEL))
        assertTrue(types.contains(ProjectEntityType.VALIDATION_STAMP))
        assertFalse(types.contains(ProjectEntityType.BUILD))
        assertFalse(types.contains(ProjectEntityType.PROMOTION_RUN))
        assertFalse(types.contains(ProjectEntityType.VALIDATION_RUN))
    }

    @Test
    fun canEdit() {
        val p1 = of(NameDescription("P1", "Project 1")).withId(of(1))
        val p2 = of(NameDescription("P2", "Project 2")).withId(of(2))
        val b1 = of(p1, NameDescription("B1", "Branch 1")).withId(of(10))
        val b2 = of(p2, NameDescription("B2", "Branch 2")).withId(of(20))

        every { securityService.isProjectFunctionGranted(1, ProjectConfig::class.java) } returns true
        every { securityService.isProjectFunctionGranted(2, ProjectConfig::class.java) } returns false

        assertTrue(type.canEdit(b1, securityService))
        assertFalse(type.canEdit(b2, securityService))
    }

    @Test
    fun canView() {
        val p1 = of(NameDescription("P1", "Project 1")).withId(of(1))
        val p2 = of(NameDescription("P2", "Project 2")).withId(of(2))
        val b1 = of(p1, NameDescription("B1", "Branch 1")).withId(of(10))
        val b2 = of(p2, NameDescription("B2", "Branch 2")).withId(of(20))
        assertTrue(type.canView(b1, securityService))
        assertTrue(type.canView(b2, securityService))
    }

    @Test
    fun forStorage() {
        val configuration = JenkinsConfiguration(
            "MyConfig",
            "http://jenkins",
            "user",
            "secret"
        )
        TestUtils.assertJsonEquals(
            mapOf(
                "configuration" to "MyConfig",
                "job" to "MyJob",
            ).asJson(),
            type.forStorage(
                JenkinsJobProperty(
                    configuration,
                    "MyJob"
                )
            )
        )
    }

    @Test
    fun fromStorage() {
        val configuration = JenkinsConfiguration(
            "MyConfig",
            "http://jenkins",
            "user",
            "secret"
        )
        every { configurationService.getConfiguration("MyConfig") } returns configuration
        // Stored JSON
        val node: JsonNode = mapOf(
            "configuration" to "MyConfig",
            "job" to "MyJob",
        ).asJson()
        // Reading
        val retrieved = type.fromStorage(node)
        assertEquals(configuration.url, retrieved.configuration.url)
        assertEquals("MyJob", retrieved.job)
    }
}
